src/Aspire.Hosting/Dashboard/GenericResourceSnapshot.cs

```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+using Google.Protobuf.WellKnownTypes;
+
+namespace Aspire.Hosting.Dashboard;
+
+internal class GenericResourceSnapshot(IResource resource, Dictionary<string, string> properties) : ResourceSnapshot
+{
+    // Default to the resource type name without the "Resource" suffix.
+    public override string ResourceType => resource.GetType().Name.Replace("Resource", "");
+
+    protected override IEnumerable<(string Key, Value Value)> GetProperties()
+    {
+        foreach (var property in properties)
+        {
+            yield return (property.Key, Value.ForString(property.Value));
+        }
+
+        if (resource is IResourceWithConnectionString connectionString)
+        {
+            var cs = connectionString.GetConnectionString();
+            yield return ("ConnectionString", cs is null ? Value.ForString(cs) : Value.ForNull());
```

Comment:
We need to be able to mark values as secret @drewnoakes

Reply:
It should be straightforward to extend the protocol for this. You can add it to the objects involved in this layer now if you like, and we can plumb that through later.

We could differentiate between secrets that require permissions and secrets that just require making in the UI by default.

Reply:
Filed https://github.com/dotnet/aspire/issues/2415